effectful/ops/types.py

```diff
@@ -15,40 +24,349 @@ class NotHandled(Exception):
+
+        return op  # type: ignore[return-value]
+
+    @define.register(type)
+    @define.register(typing.cast(type, types.GenericAlias))
+    @define.register(typing.cast(type, typing._GenericAlias))  # type: ignore[attr-defined]
+    @define.register(typing.cast(type, types.UnionType))
+    @classmethod
+    def _define_type[T](
+        cls, t: type[T], *, name: str | None = None
+    ) -> "Operation[[], T]":
+        def func() -> t:  # type: ignore
+            raise NotHandled
+
+        typing.get_type_hints(func)
+
+        if name is None:
+            name = t.__name__
+
+        return typing.cast(Operation[[], T], cls.define(func, name=name))
+
+    @define.register(types.BuiltinFunctionType)
+    @classmethod
+    def _define_builtinfunctiontype[**P, T](
+        cls, t: Callable[P, T], *, name: str | None = None
+    ) -> "Operation[P, T]":
+        @functools.wraps(t)
+        def func(*args, **kwargs):
+            from effectful.ops.semantics import fvsof
+
+            if not fvsof((args, kwargs)):
+                return t(*args, **kwargs)
+            else:
+                raise NotHandled
+
+        return typing.cast(Operation[P, T], cls.define(func, name=name))
+
+    @define.register(staticmethod)
+    @classmethod
+    def _define_staticmethod[**P, T](cls, t: "staticmethod[P, T]", **kwargs):
+        return staticmethod(cls.define(t.__func__, **kwargs))
+
+    @define.register(functools.singledispatchmethod)
+    @classmethod
+    def _define_singledispatchmethod(cls, default, **kwargs):
+        if isinstance(default.func, classmethod):
+            raise NotImplementedError("Operations as classmethod are not yet supported")
+
+        @functools.wraps(default.func)
+        def _wrapper(obj, *args, **kwargs):
+            return default.__get__(obj)(*args, **kwargs)
+
+        op = cls.define(_wrapper, **kwargs)
+        op.register = default.register
+        op.__isabstractmethod__ = default.__isabstractmethod__
+        return op
+
+    @typing.runtime_checkable
+    class _SingleDispatchCallable(typing.Protocol):
+        registry: types.MappingProxyType[object, Callable]
+
+        def dispatch(self, cls: type) -> Callable: ...
+        def register(self, cls: type, func: Callable | None = None) -> Callable: ...
+        def _clear_cache(self) -> None: ...
+        def __call__(self, /, *args, **kwargs): ...
+
+    @define.register(_SingleDispatchCallable)
+    @classmethod
+    def _defop_singledispatchoperation(cls, default, **kwargs):
+        op = cls._define_callable(default, **kwargs)
+        op.dispatch = default._registry.dispatch
+        op.register = default._registry.register
+        return op
+
+    @define.register(_CustomSingleDispatchCallable)
+    @classmethod
+    def _defop_customsingledispatchcallable(cls, default, **kwargs):
+        sig = inspect.signature(functools.partial(default.func, None))
+        op = cls._define_callable(default, signature=sig)
```

Comment:
ditto: we should only be accessing `_define_callable` indirectly via type dispatch